plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/layout/provider/BorderItemAwareLayoutProvider.java

```diff
@@ -513,10 +506,22 @@ public Command layoutEditParts(final List selectedObjects, final IAdaptable layo
 
         // Finds if there are unpinned diagram elements to keep fixed stored in
         // the LayoutHint as a Collection
-        ArrayList<IDiagramElementEditPart> elementsToKeepFixed = new ArrayList<>();
-        if (layoutHint.getAdapter(Collection.class) instanceof ArrayList<?>
-                && Iterables.all((ArrayList<?>) layoutHint.getAdapter(Collection.class), validateAllElementInArrayListAreIDiagramElementEditPart)) {
-            elementsToKeepFixed = (ArrayList<IDiagramElementEditPart>) layoutHint.getAdapter(Collection.class);
+        Set<IGraphicalEditPart> elementsToKeepFixed = new HashSet<>();
+        Collection<EditPart> adaptedCollection = layoutHint.getAdapter(Collection.class);
+        if (adaptedCollection != null && adaptedCollection.stream().allMatch(IDiagramElementEditPart.class::isInstance)) {
+            elementsToKeepFixed.addAll(adaptedCollection.stream().map(IDiagramElementEditPart.class::cast).toList());
+        }
+        if (SiriusLayoutDataManager.KEEP_FIXED.equals(layoutHint.getAdapter(String.class))) {
+            elementsToKeepFixed.addAll(selectedObjects.stream().filter(IGraphicalEditPart.class::isInstance).toList());
+        }
+
+        Set<DDiagramElement> temporaryPinnedElements = new HashSet<>();
+        PinHelper pinHelper = new PinHelper();
+        for (IGraphicalEditPart elementEP : elementsToKeepFixed) {
+            if (elementEP.resolveSemanticElement() instanceof DDiagramElement dDiagramElement && !pinHelper.isPinned(dDiagramElement)) {
+                pinHelper.markAsPinned(dDiagramElement);
+                temporaryPinnedElements.add(dDiagramElement);
+            }
```

Comment:
Check that it works in collaborative mode. Also, as you said, you'd have to check that this doesn't pose a problem with the "Move Pinned Elements" feature.

Reply:
OK it works in collab
TODO: check with the "Move Pinned Elements" feature

Reply:
It seems to work with "Move Pinned Elements" feature and also with "Layout Linked Border Node of the Selection", I think it's because the `elementsToKeepFixed` are kept in addition to the pinning. So it should be fine for that.